File: packages/browser/src/registry.ts
import { BrowserConfig, type BrowserBackendKind, type BrowserConfig as BrowserConfigType } from "./schemas.js";
import type { BrowserProvider, CreateSessionResult } from "./provider.js";

export interface ResolvedBackend {
  kind: BrowserBackendKind;
  provider: BrowserProvider;
  reason: string;
}

export interface RegistryOptions {
  /** Override precedence: env/config CDP URL beats everything else when set. */
  cdpUrl?: string;
  /** Explicitly configured backend/provider key. */
  configuredBackend?: BrowserBackendKind;
  /** Optional environment-based availability hints. */
  env?: Record<string, string | undefined>;
}

export class BrowserProviderRegistry {
  private readonly providers = new Map<BrowserBackendKind, BrowserProvider>();

  register(provider: BrowserProvider): void {
    this.providers.set(provider.name, provider);
  }

  get(kind: BrowserBackendKind): BrowserProvider | undefined {
    return this.providers.get(kind);
  }

  list(): BrowserProvider[] {
    return Array.from(this.providers.values());
  }

  /**
   * Resolve the active backend following the Core precedence:
   *   1. explicit CDP URL override -> cdp backend
   *   2. explicit configured backend (if available)
   *   3. legacy preference walk: browser-use -> browserbase -> camofox -> local
   *   4. local fallback
   */
  async resolve(config: unknown, options?: RegistryOptions): Promise<ResolvedBackend> {
    const parsed = BrowserConfig.parse(config) as BrowserConfigType;
    const env = options?.env ?? {};

    // 1. CDP URL override
    const cdpUrl = options?.cdpUrl ?? parsed.cdpUrl ?? env.BROWSER_CDP_URL;
    if (cdpUrl) {
      const provider = this.providers.get("cdp");
      if (provider) {
        return { kind: "cdp", provider, reason: "cdp_url_override" };
      }
    }

    // 2. Explicit configured backend
    const configured = options?.configuredBackend ?? parsed.backend;
    if (configured !== "local") {
      const provider = this.providers.get(configured);
      if (provider && (await provider.isAvailable(parsed))) {
        return { kind: configured, provider, reason: "configured_backend" };
      }
    }

    // 3. Legacy preference walk (mirror agent/browser_registry.py)
    const legacyOrder: BrowserBackendKind[] = ["browser-use", "browserbase", "camofox"];
    for (const kind of legacyOrder) {
      const provider = this.providers.get(kind);
      if (!provider) continue;
      if (await provider.isAvailable(parsed)) {
        return { kind, provider, reason: "legacy_preference" };
      }
    }

    // 4. local fallback
    const local = this.providers.get("local");
    if (local) {
      return { kind: "local", provider: local, reason: "local_fallback" };
    }

    throw new Error("No browser provider is available");
  }
}

/**
 * Shared registry instance used by the desktop agent loop.
 */
export const browserRegistry = new BrowserProviderRegistry();

export type { BrowserProvider, CreateSessionResult };

File: src/browser/registry.rs
//! Backend precedence algorithm (mirrors `packages/browser/src/registry.ts`).
//!
//! The TS `BrowserProviderRegistry` owns the runtime provider objects; this pure
//! function encodes only the 4-step precedence so it can be tested independently
//! and reused as a single source of truth in the future Rust sidecar.

use crate::browser::types::BrowserBackendKind;

/// Result of resolving the active backend.
#[derive(Debug, Clone, PartialEq)]
pub struct ResolvedBackend {
    pub kind: BrowserBackendKind,
    pub reason: String,
}

/// Resolve the active browser backend following the Core precedence:
///   1. explicit CDP URL override -> cdp backend
///   2. explicit configured backend (if available and not local)
///   3. legacy preference walk: browser-use -> browserbase -> camofox
///   4. local fallback
///
/// `config_backend` is the config's `backend`; `cdp_url` is the *effective* CDP
/// URL (already combined from options/config/env); `configured_backend` is the
/// optional explicit override; `available` reports provider availability.
pub fn resolve_backend(
    config_backend: BrowserBackendKind,
    cdp_url: Option<&str>,
    configured_backend: Option<BrowserBackendKind>,
    available: &dyn Fn(BrowserBackendKind) -> bool,
) -> Result<ResolvedBackend, String> {
    // 1. CDP URL override
    if cdp_url.is_some() && available(BrowserBackendKind::Cdp) {
        return Ok(ResolvedBackend {
            kind: BrowserBackendKind::Cdp,
            reason: "cdp_url_override".to_string(),
        });
    }

    // 2. Explicit configured backend (if available and not local)
    let configured = configured_backend.unwrap_or(config_backend);
    if configured != BrowserBackendKind::Local && available(configured) {
        return Ok(ResolvedBackend {
            kind: configured,
            reason: "configured_backend".to_string(),
        });
    }

    // 3. Legacy preference walk
    for kind in [
        BrowserBackendKind::BrowserUse,
        BrowserBackendKind::Browserbase,
        BrowserBackendKind::Camofox,
    ] {
        if available(kind) {
            return Ok(ResolvedBackend {
                kind,
                reason: "legacy_preference".to_string(),
            });
        }
    }

    // 4. Local fallback
    if available(BrowserBackendKind::Local) {
        return Ok(ResolvedBackend {
            kind: BrowserBackendKind::Local,
            reason: "local_fallback".to_string(),
        });
    }

    Err("No browser provider is available".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    // Build an availability predicate from a set of always-available kinds.
    fn available_from(kinds: &[BrowserBackendKind]) -> impl Fn(BrowserBackendKind) -> bool {
        let set = kinds.iter().copied().collect::<Vec<_>>();
        move |kind| set.contains(&kind)
    }

    #[test]
    fn prefers_cdp_url_override_over_configured_backend() {
        let available = available_from(&[
            BrowserBackendKind::Cdp,
            BrowserBackendKind::Browserbase,
            BrowserBackendKind::Local,
        ]);
        let resolved = resolve_backend(
            BrowserBackendKind::Browserbase,
            Some("ws://127.0.0.1:9222/devtools/browser/abc"),
            None,
            &available,
        )
        .unwrap();
        assert_eq!(resolved.kind, BrowserBackendKind::Cdp);
        assert_eq!(resolved.reason, "cdp_url_override");
    }

    #[test]
    fn uses_explicit_configured_backend_when_available() {
        let available =
            available_from(&[BrowserBackendKind::Browserbase, BrowserBackendKind::Local]);
        let resolved =
            resolve_backend(BrowserBackendKind::Browserbase, None, None, &available).unwrap();
        assert_eq!(resolved.kind, BrowserBackendKind::Browserbase);
        assert_eq!(resolved.reason, "configured_backend");
    }

    #[test]
    fn falls_back_through_legacy_preference_order() {
        let available =
            available_from(&[BrowserBackendKind::BrowserUse, BrowserBackendKind::Local]);
        let resolved = resolve_backend(BrowserBackendKind::Local, None, None, &available).unwrap();
        assert_eq!(resolved.kind, BrowserBackendKind::BrowserUse);
        assert_eq!(resolved.reason, "legacy_preference");
    }

    #[test]
    fn falls_back_to_local_when_nothing_else_is_available() {
        let available = available_from(&[BrowserBackendKind::Local]);
        let resolved =
            resolve_backend(BrowserBackendKind::Browserbase, None, None, &available).unwrap();
        assert_eq!(resolved.kind, BrowserBackendKind::Local);
        assert_eq!(resolved.reason, "local_fallback");
    }

    #[test]
    fn throws_when_no_provider_is_available() {
        let available = available_from(&[]);
        let err = resolve_backend(BrowserBackendKind::Local, None, None, &available).unwrap_err();
        assert_eq!(err, "No browser provider is available");
    }

    #[test]
    fn env_hints_influence_availability() {
        let available =
            available_from(&[BrowserBackendKind::Browserbase, BrowserBackendKind::Local]);
        let resolved =
            resolve_backend(BrowserBackendKind::Browserbase, None, None, &available).unwrap();
        assert_eq!(resolved.kind, BrowserBackendKind::Browserbase);
        assert_eq!(resolved.reason, "configured_backend");
    }
}
